### PR TITLE
[X86] Change target of __builtin_ia32_cmp[p|s][s|d] from avx into sse/sse2

### PR DESCRIPTION
**Pull request by FreddyLeaf**

### CODE REVIEW

**Comment by FreddyLeaf** [comment]
This PR is to align with icc: https://godbolt.org/z/EzbfzTrzr We can modify intrinsic guide if required.

intel.com/SDM shows:
```
CMPSD xmm1, xmm2/m64, imm8
SSE2: Compare low double-precision floating-point value in xmm2/m64 and xmm1 using bits 2:0 of imm8 as comparison predicate
Intel C/C++ Compiler Intrinsic Equivalent
(V)CMPSD _m128d _mm_cmp_sd(_m128d a, __m128d b, const int imm)

File: clang/lib/Headers/emmintrin.h [hunk 1]
```diff
@@ -4742,6 +4742,125 @@ static __inline__ __m128d __DEFAULT_FN_ATTRS _mm_castsi128_pd(__m128i __a) {
...
+///    the two comparison results: zero if the comparison is false, and all 1's
+///    if the comparison is true.
+///
+/// \headerfile <x86intrin.h>
+///
+/// \code
+/// __m128d _mm_cmp_pd(__m128d a, __m128d b, const int c);
+/// \endcode
+///
+/// This intrinsic corresponds to the <c> VCMPPD </c> instruction.
```
Comment (RKSimon) [inline clang/lib/Headers/emmintrin.h:4759]: (V)CMPPD ?

File: clang/lib/Headers/emmintrin.h [hunk 2]
```diff
@@ -4742,6 +4742,125 @@ static __inline__ __m128d __DEFAULT_FN_ATTRS _mm_castsi128_pd(__m128i __a) {
...
+///
+/// \param a
+///    A 128-bit vector of [2 x double].
+/// \param b
+///    A 128-bit vector of [2 x double].
+/// \param c
+///    An immediate integer operand, with bits [4:0] specifying which comparison
```
Comment (RKSimon) [inline clang/lib/Headers/emmintrin.h:4766]: What should we do about non-AVX targets only supporting 3-bit imm modes (vs AVX 5-bit imm nodes)?

**Review by RKSimon — COMMENTED** [review]
No strong objections, but I think the documentation needs to be tweaked to better explain SSE vs AVX handling.

**Comment by FreddyLeaf** [comment]
> No strong objections, but I think the documentation needs to be tweaked to better explain SSE vs AVX handling.

I have an idea to show two results for the `_mm_cmp_sd`, one's CPUID is SSE2, the other one's is AVX.

File: clang/lib/Headers/emmintrin.h (hunk 2 above)
Comment (phoebewang) [reply clang/lib/Headers/emmintrin.h:4766]: How about add sema check for SSE case? Users need to remind they may get unexpected result due to not enable AVX feature.

Comment (FreddyLeaf) [reply clang/lib/Headers/emmintrin.h:4766]:
```
$ echo "cmpsd   \$8, %xmm1, %xmm0" | llvm-mc --show-encoding
        cmpsd   $8, %xmm1, %xmm0                # encoding: [0xf2,0x0f,0xc2,0xc1,0x08]
$ echo "cmpsd   \$0, %xmm1, %xmm0" | llvm-mc --show-encoding
        cmpeqsd %xmm1, %xmm0                    # encoding: [0xf2,0x0f,0xc2,0xc1,0x00]
```
encoding won't omit higher bits for cmpsd, but maybe hardware will?

Comment (FreddyLeaf) [reply clang/lib/Headers/emmintrin.h:4766]: > How about add sema check for SSE case? Users need to remind they may get unexpected result due to not enable AVX feature.

Good idea, addressed in 85d84aa16d142f3541390307184cd95cdf548813

**Review by phoebewang — APPROVED** [review]
LGTM.

File: clang/lib/Headers/xmmintrin.h [hunk 3]
```diff
@@ -2936,6 +2936,162 @@ _mm_movemask_ps(__m128 __a)
...
+#define _CMP_LE_OS 0x02   /* Less-than-or-equal (ordered, signaling)  */
+#define _CMP_UNORD_Q 0x03 /* Unordered (non-signaling)  */
+#define _CMP_NEQ_UQ 0x04  /* Not-equal (unordered, non-signaling)  */
+#define _CMP_NLT_US 0x05  /* Not-less-than (unordered, signaling)  */
+#define _CMP_NLE_US 0x06  /* Not-less-than-or-equal (unordered, signaling)  */
+#define _CMP_ORD_Q 0x07   /* Ordered (non-signaling)   */
+#define _CMP_EQ_UQ 0x08   /* Equal (unordered, non-signaling)  */
+#define _CMP_NGE_US                                                            \
+  0x09                     /* Not-greater-than-or-equal (unordered, signaling) \
+                            */
```
Comment (RKSimon) [inline clang/lib/Headers/xmmintrin.h:2951]: keep to a single line - avoid clang-format in these kinds of cases

**Review by RKSimon — COMMENTED** [review]
Are we still missing SSE vs AVX sema checking test coverage?

File: clang/lib/Headers/xmmintrin.h (hunk 3 above)
Comment (FreddyLeaf) [reply clang/lib/Headers/xmmintrin.h]: 96ca4d04529e70d8ec75e7f44ddd35f072fe4d10

**Comment by FreddyLeaf** [comment]
> Are we still missing SSE vs AVX sema checking test coverage?

yes, added in 96ca4d04529e70d8ec75e7f44ddd35f072fe4d10

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by bgra8** [comment]
@FreddyLeaf functions with `__attribute__((target("avx")))` trigger the newly added warning.

Code like this no longer builds and creates a lot of fallout for us (google internal code):

```
#include <immintrin.h>

__attribute__((target("avx"))) void test(__m128 a, __m128 b) {
  _mm_cmp_ps(a, b, 14);
}
```

Compiler output:
```
test.cc:4:3: error: argument value 14 is outside the valid range [0, 7] [-Wargument-outside-range]
    4 |   _mm_cmp_ps(a, b, 14);
```

Passing `-mavx` in the compilation command line makes the build green but the attribute `target("avx")` should be equivalent.

Please revert

**Pull request merged**